### ISSUE TITLE
How about separating packaging(nuget)?

### ISSUE THREAD

**Issue by IOL0ol1** [Mapsui/Mapsui#1034]
I don't think it's a normal practice to make an all-in-one nuget package.
Is there any special purpose in the current packaging method ?
I forked a branch and tried to separate the dll packaging, while adding support for the lower net framework(net462).
I released the [mapsui3 series of nuget packages](https://www.nuget.org/packages?q=mapsui3) for my  wpf(net47) project.
We should select the nuget package for the corresponding platform without having to install it all.

**Comment by pauldendulk** [comment]
Each dll in a separate package was not the best practice when Mapsui started. Changing it will be a breaking change, but could be acceptable. We can remove all the nuspecs and will not have to update them (I frequently forget about that). So perhaps we should go for it. 

Anyone else objections? @janusw What do you think?

**Comment by janusw** [comment]
@IOL0ol1 What would be the major advantage that you gain from breaking up Mapsui in several packages? What's your motivation for doing that?

IMHO your breakup into
* Mapsui
* Mapsui.Geometries
* Mapsui.Rendering.Xaml
* Mapsui.Rendering.Skia
* Mapsui.Desktop
* Mapsui.UI.Wpf

is a bit of an overkill, also it's not complete (you're missing things like Mapsui.UI.Android, Mapsui.UI.iOS, etc), right?

@pauldendulk  In my opinion Mapsui may need a bit of restructuring first, before breaking up the package. For example, many of the classes that are included in the Mapsui.Desktop VS project actually live in namespaces like `Mapsui.Providers`. That should be cleaned up first. Possibly Mapsui.Desktop can be removed completely? Looks pretty much like a collection of legacy code to me ...

**Comment by janusw** [comment]
> @pauldendulk In my opinion Mapsui may need a bit of restructuring first, before breaking up the package. For example, many of the classes that are included in the Mapsui.Desktop VS project actually live in namespaces like `Mapsui.Providers`. That should be cleaned up first. Possibly Mapsui.Desktop can be removed completely? Looks pretty much like a collection of legacy code to me ...

An example of what I'm talking about here is the `WFSProvider` and related code. AFAICS its namespace suggests it should go into the Mapsui project: https://github.com/janusw/Mapsui/commits/wfs

I don't see what is 'Desktop'-specific about it (possibly the same for WmsProvider etc). If you don't want them in the Mapsui core project, maybe add a new project 'Mapsui.OGCProviders' or 'Mapsui.ExtendedProviders'. Mapsui.Desktop is somewhat misnamed AFAICS.

**Comment by pauldendulk** [comment]
At some point Mapsui started cross platform support, using PCLs in those days, but some parts could not be ported to PCLs but were still in use in desktop apps. From what I remember the limitations were in:
- Xml. It used an older library. This should be portable.
- Web. Just now looking at how it is used and it seems to make not sense.
- File. I think this is now all part of .netstandard, but perhaps not fully implemented.

Perhaps it is now time to port those parts or remove them.

btw, are you working on a Mac?

**Comment by pauldendulk** [comment]
Just removed System.Web and irrelevant web caching from Shapefile
https://github.com/Mapsui/Mapsui/pull/1069

**Comment by janusw** [comment]
> At some point Mapsui started cross platform support, using PCLs in those days, but some parts could not be ported to PCLs but were still in use in desktop apps. From what I remember the limitations were in:
> 
>     [...]
>
> Perhaps it is now time to port those parts or remove them.

Agree :)

If there are still any obstacles now, one might have to wait until .NET 6 is there.

I'd be particularly interested in making the WFS part work. May look into that soon ...


> btw, are you working on a Mac?

I am (most of the time) ...

**Comment by pauldendulk** [comment]
Moved Shapefile to the Mapsui project. 
https://github.com/Mapsui/Mapsui/pull/1070

**Comment by pauldendulk** [comment]
@janusw Just looked at WFS. When this is moved to Mapsui there are no compile errors but I cannot check if it is functioning correctly because there are no working samples. One way would be to restore the geoserver sample but this is not something I would like to dig into. We need some open WFS service we could use as sample. 

The WFS originates from SharpMap, do they have a better sample there? Do you have experience working with WFS with SharpMap?

**Comment by JBou** [comment]
I was working on fixing the WFS Provider for the last two weeks. I merged some code from SharpMap, the Mapsui WFS Provider was really outdated and had different bugs. There are still some things missing, though.

I will create a PR later today. There is also a working sample where you can click on the map and get a highlighted shape and information on that specific point. The one thing missing is Axis Order in higher WFS versions. Was looking into it yesterday, it's still not clear if it is working now and needs some testing with different CRS's. Would like to discuss my changes on the PR later on.

**Comment by janusw** [comment]
> @janusw Just looked at WFS. When this is moved to Mapsui there are no compile errors but I cannot check if it is functioning correctly because there are no working samples.

Same here. I will make a PR out of my branch which does this.

> One way would be to restore the geoserver sample but this is not something I would like to dig into. We need some open WFS service we could use as sample.

I know a few open WFS examples, which I will post soon. We should move this discussion into issue #1072, which I have just opened for this.

 
> The WFS originates from SharpMap, do they have a better sample there? Do you have experience working with WFS with SharpMap?

No, I have never used SharpMap.

**Comment by janusw** [comment]
> I was working on fixing the WFS Provider for the last two weeks. I merged some code from SharpMap, the Mapsui WFS Provider was really outdated and had different bugs. There are still some things missing, though.
> 
> I will create a PR later today. There is also a working sample where you can click on the map and get a highlighted shape and information on that specific point. The one thing missing is Axis Order in higher WFS versions. Was looking into it yesterday, it's still not clear if it is working now and needs some testing with different CRS's. Would like to discuss my changes on the PR later on.

Sounds great! Any help is appreciated here. Let's move the discussion into #1072.

**Comment by janusw** [comment]
> > @janusw Just looked at WFS. When this is moved to Mapsui there are no compile errors but I cannot check if it is functioning correctly because there are no working samples.
> 
> Same here. I will make a PR out of my branch which does this.

See #1073.

**Comment by janusw** [comment]
Coming back to the core topic of this issue ...

> @IOL0ol1 What would be the major advantage that you gain from breaking up Mapsui in several packages? What's your motivation for doing that?
> 
> IMHO your breakup into
> 
>     * Mapsui
> 
>     * Mapsui.Geometries
> 
>     * Mapsui.Rendering.Xaml
> 
>     * Mapsui.Rendering.Skia
> 
>     * Mapsui.Desktop
> 
>     * Mapsui.UI.Wpf
> 
> 
> is a bit of an overkill, also it's not complete (you're missing things like Mapsui.UI.Android, Mapsui.UI.iOS, etc), right?

One scheme that would actually make sense to me would be to split off all Mapsui.UI.* parts from the Mapsui base package, so that Mapsui is separated into three packages:
1. the `Mapsui` base pkg (including Mapsui.Geometries and Mapsui.Rendering.Skia)
2. `Mapsui.UI` (or `Mapsui.Native`? or `Mapsui.UI.Native`?), containing: Mapsui.UI.Android, Mapsui.UI.iOS, Mapsui.UI.Uwp, Mapsui.UI.Wpf, Mapsui.Rendering.Xaml
3. `Mapsui.Forms` (or `Mapsui.UI.Forms`?)

Why does this make sense? Because people usually use either Mapsui.UI.Forms or Mapsui.UI.Native, but not both of the same time. Everyone needs the Mapsui base package, therefore both UI packages will depend on it.

Further advantages:
* Both Mapsui and Mapsui.UI.Forms can be built on any platform this way (Win/Mac/Linux). Only Mapsui.UI.Native must be built on Windows.
* Bugs in the Mapsui.UI.Native part will not affect Mapsui.UI.Forms (e.g. #1115).

Possibly one could further split up Mapsui.UI.Native into two distinct packages:
* Mapsui.UI.Win (with UWP & WPF)
* Mapsui.UI.Mobile (with Android & iOS)

However, this is not really essential to me, and could be deferred to a later stage.

@pauldendulk, what do you think about this scheme?

**Comment by IOL0ol1** [comment]
@janusw I'm here. I found that separating the nuget package required changes to the project structure.Therefore, I do not continue to pay attention to this issue.

**Comment by IOL0ol1** [comment]
Mapsui.Core

**Comment by janusw** [comment]
> I found that separating the nuget package required changes to the project structure.

Well, I guess this is not strictly required, but it does make sense to have the project structure match the packaging (and the namespacing). Both does not fully match right now, but there have been a few changes already, and some are still outstanding (e.g. #1165).

> Therefore, I do not continue to pay attention to this issue.

Don't worry, your input is still appreciated. But to pull it through, it helps to have feedback from someone with a good overview over Mapsui. @pauldendulk as lead developer is obviously someone like this. Maybe also @charlenni has some feedback on the scheme I proposed above?

@IOL0ol1, do you have an opinion on what I proposed? Would it work well with your use cases?

**Comment by IOL0ol1** [comment]
@janusw 
In general, different platforms use different nuget packages.

- Mapsui.Core (the Mapsui base pkg)
- Mapsui.UI.Wpf 
- Mapsui.UI.Uwp 
- Mapsui.UI.Android 
- Mapsui.UI.iOS 
- Mapsui   (No code, only .target script, detect different platforms, and install the nuget package above)      

maybe 
- Mapsui.XXXX.Extension  (some extension method with third-party libraries)


Visual Studio 2019 provides the following three project templates:    
1. Mobile app(Xamarin.Forms)  support iOS and Android use Xamarin.Forms's .xaml for UI layout    
2. iOS app(Xamarin)  support iOS use iOS's .storyboard for UI layout  
3. Android app(Xamarin) support Android use android's .xml for UI layout  

It depends on the type of project: Mapsui.UI.Android , Mapsui.UI.iOS or just Mapsui.UI.Mobile.

Mapsui.Geometries move to Mapsui/Geometries(Mapsui.Core)

System.Drawing.Common support MonoAndroid,iOS,Mac,Windows,Linux now,    
so maybe Mapsui.Desktop can be merged into Mapsui/Desktop(Mapsui.Core)     
or as an extension package (use extension method).    

It's a bit of a hassle to modify Mapsui.Rendering.Xaml.    
Maybe xaml and skia should only choose one.    
Installing different packages corresponds to different implementations.   
Of course, it might be a good idea for us to just give up the xaml renderer,    
merged skia to Mapsui/Rendering(Mapsui.Core).

**Comment by janusw** [comment]
> @janusw
> In general, different platforms use different nuget packages.
> 
>     * Mapsui.Core (the Mapsui base pkg)
>     * Mapsui.UI.Wpf
>     * Mapsui.UI.Uwp
>     * Mapsui.UI.Android
>     * Mapsui.UI.iOS

Well, yeah, in principle one can do that. But what's the major argument for this? Package size reduction, I guess? (Since one can just include one particular platform and leave out the stuff that's needed for the others.) Anything else?

For me the most important point is separating the UI.Forms part from the rest of the UI platforms. (There is already a separate Mapsui.Forms package, but it depends on the Mapsui base pkg, which includes all the other 'native' UI stuff).

In any case, I'd consider the Mapsui.UI.Android and Android.UI.iOS as 'legacy', since anyone targeting mobile platforms would rather use Mapsui.UI.Forms now, I guess.

>     * Mapsui   (No code, only .target script, detect different platforms, and install the nuget package above)
> 
> maybe
> 
>     * Mapsui.XXXX.Extension  (some extension method with third-party libraries)

None of these two are really necessary IMHO.


> Visual Studio 2019 provides the following three project templates:
> 
>     1. Mobile app(Xamarin.Forms)  support iOS and Android use Xamarin.Forms's .xaml for UI layout
> 
>     2. iOS app(Xamarin)  support iOS use iOS's .storyboard for UI layout
> 
>     3. Android app(Xamarin) support Android use android's .xml for UI layout
> 
> 
> It depends on the type of project: Mapsui.UI.Android , Mapsui.UI.iOS or just Mapsui.UI.Mobile.

The first one would use Mapsui.UI.Forms, of course.

 
> Mapsui.Geometries move to Mapsui/Geometries(Mapsui.Core)

Agreed. I have opened #1172 for this.
 
> System.Drawing.Common support MonoAndroid,iOS,Mac,Windows,Linux now,
> so maybe Mapsui.Desktop can be merged into Mapsui/Desktop(Mapsui.Core)
> or as an extension package (use extension method).

Yes. See #1165.

> It's a bit of a hassle to modify Mapsui.Rendering.Xaml.
> Maybe xaml and skia should only choose one.
> Installing different packages corresponds to different implementations.
> Of course, it might be a good idea for us to just give up the xaml renderer,
> merged skia to Mapsui/Rendering(Mapsui.Core).

Well, yeah. The xaml renderer is only for WPF as far as I know (never used it). In the long term it should be removed, I guess. But there was some issue with skia on WPF which is blocking this IIRC (but I don't remember what it was).

**Comment by janusw** [comment]
> For me the most important point is separating the UI.Forms part from the rest of the UI platforms. (There is already a separate Mapsui.Forms package, but it depends on the Mapsui base pkg, which includes all the other 'native' UI stuff).

This has been implemented with #1175, which has been merged by now.

**Comment by janusw** [comment]
I think this issue can be closed as well, after #1207 is done. The new packages are available with today's release 3.0.0-rc.3.
